### DISCUSSION TITLE
Model's format convertion

### DISCUSSION THREAD

**Discussion by max-dev-14** [ultralytics/ultralytics#21720] [Q&A] [answered]
Greeting! How can I convert an ONNX-formatted model (also hand-made one) to a PyTorch-formatted one for further working using Ultralytics framework?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @max-dev-14, thank you for your interest in Ultralytics 🚀! This is an automated response — an Ultralytics engineer will also assist you here soon.

We recommend a visit to the Docs for new users where you can find many Python and CLI usage examples and where many of the most common questions may already be answered:
- Docs: https://docs.ultralytics.com/
- Python: https://docs.ultralytics.com/usage/python/
- CLI: https://docs.ultralytics.com/usage/cli/

If this is a 🐛 Bug Report, please provide a minimum reproducible example to help us debug it:
- MRE: https://docs.ultralytics.com/help/minimum-reproducible-example/

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our Tips for Best Training Results:
- Training Tips: https://docs.ultralytics.com/guides/model-training-tips/

Join the Ultralytics community where it suits you best:
- Discord (real-time chat) 🎧: https://discord.com/invite/ultralytics
- Discourse (in-depth discussions): https://community.ultralytics.com/
- Subreddit (community threads): https://www.reddit.com/r/Ultralytics/

## Upgrade
Upgrade to the latest ultralytics package including all requirements in a Python>=3.8 environment with PyTorch>=1.8 to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
- Requirements: https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml

## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):
- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See GCP Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/
- Amazon Deep Learning AMI. See AWS Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/
- Docker Image. See Docker Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/ <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all Ultralytics CI tests are currently passing. CI tests verify correct operation of all YOLO Modes and Tasks on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply, marked as answer]
Hi @max-dev-14 — converting ONNX back to a PyTorch .pt model isn’t supported or reliable, and Ultralytics doesn’t provide an ONNX→PyTorch tool. You can load and run ONNX directly in Ultralytics for predict/val/benchmark as shown in the ONNX integration guide, but for training/finetuning you’ll need the original PyTorch model/weights or retrain. If your ONNX is from another repo, you may need to adapt its outputs for compatibility; see the YOLOv7 ONNX compatibility steps for an example.

**Comment by max-dev-14** [comment]
Many thanks

**Reply by glenn-jocher** [reply]
You're welcome—if you decide to run the ONNX directly in Ultralytics, you can do so like this:
from ultralytics import YOLO
model = YOLO("your_model.onnx", task="detect")
results = model("path/to/image.jpg")
